### PR TITLE
[refactor] Predicate push down

### PR DESCRIPTION
**Pull request by systay**
## Description
This is mostly a refactoring of how predicates are pushed down the operator tree. The exception is how we handle predicate pushing into `UNION` clauses. This can happen if we have a predicate on a derived table that consists of an `UNION`. The old behaviour was not really correct.

## Checklist
-   [x] "Backport me!" label has been added if this change should be backported
-   [x] Tests were added or are not required
-   [x] Documentation was added or is not required

### CODE REVIEW

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 1]
```diff
@@ -73,3 +75,56 @@ func (a *ApplyJoin) Clone(inputs []Operator) Operator {
...
+	a.LHS = lhs
+	vars := map[string]int{}
+	for i, idx := range idxs {
+		vars[bvName[i]] = idx
+	}
+	rhs, err := PushPredicate(ctx, predicate, a.RHS)
+	if err != nil {
+		return err
+	}
+	a.RHS = rhs
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:127]: you could use `a.setRHS(rhs)` here

File: go/vt/vtgate/planbuilder/operators/apply_join.go [hunk 2]
```diff
@@ -73,3 +75,56 @@ func (a *ApplyJoin) Clone(inputs []Operator) Operator {
...
+func (a *ApplyJoin) addJoinPredicate(ctx *plancontext.PlanningContext, expr sqlparser.Expr) error {
+	bvName, cols, predicate, err := BreakExpressionInLHSandRHS(ctx, expr, TableID(a))
+	if err != nil {
+		return err
+	}
+	lhs, idxs, err := PushOutputColumns(ctx, a.LHS, cols...)
+	if err != nil {
+		return err
+	}
+	a.LHS = lhs
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/apply_join.go:118]: Same here, you could use `setLHS`

File: go/vt/vtgate/planbuilder/operators/apply_join.go (hunk 1 above)
Comment (systay) [reply go/vt/vtgate/planbuilder/operators/apply_join.go:127]: sure. what would be the benefit? the interface methods are useful when you can't touch the fields directly, but here I can

File: go/vt/vtgate/planbuilder/operators/operator_funcs.go [hunk 3]
```diff
@@ -368,3 +299,104 @@ func BreakExpressionInLHSandRHS(
...
+type joinOperator interface {
+	Operator
+	getLHS() Operator
+	getRHS() Operator
+	setLHS(Operator)
+	setRHS(Operator)
+	makeInner()
+	isInner() bool
+	addJoinPredicate(ctx *plancontext.PlanningContext, expr sqlparser.Expr) error
+}
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/operator_funcs.go:312]: nit: i think we should move this to a better place, like in one of the join files (`apply_join.go` or `join.go`)

File: go/vt/vtgate/planbuilder/operators/operator_funcs.go [hunk 4]
```diff
@@ -368,3 +299,104 @@ func BreakExpressionInLHSandRHS(
...
+		return join, err
+
+	case deps.IsSolvedBy(TableID(join)):
+		// if we are dealing with an outer join, always start by checking if this predicate can turn
+		// the join into an inner join
+		if !join.isInner() && canConvertToInner(ctx, expr, TableID(join.getRHS())) {
+			join.makeInner()
+		}
+
+		if !join.isInner() {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/operator_funcs.go:353]: it seems that `makeInner` always change the join into an inner join for both `applyJoin` and `join`.

File: go/vt/vtgate/planbuilder/operators/vindex.go [hunk 5]
```diff
@@ -82,11 +83,54 @@ outer:
 	return idxs, nil
 }
 
-// CheckValid implements the Operator interface
+// checkValid implements the Operator interface
 func (v *Vindex) CheckValid() error {
```
Comment (frouioui) [inline go/vt/vtgate/planbuilder/operators/vindex.go:87]: i think you can revert this change

**Review by frouioui — APPROVED** [review]
